### PR TITLE
Documentation: add readme file for ccnp-device-plugin

### PR DESCRIPTION
**Pull request by hairongchen**
Signed-off-by: Hairong Chen hairong.chen@intel.com

### CODE REVIEW

File: device-plugin/ccnp-device-plugin/README.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+```
+kubectl apply -k https://github.com/kubernetes-sigs/node-feature-discovery/deployment/overlays/default?ref=v0.14
+```
+
+before that, please build own image and deploy:
+```
+git clone https://github.com/kubernetes-sigs/node-feature-discovery.git
+cd node-feature-discovery/
+make image
+kubectl apply -k kustomization.yaml
```
Comment (Ruoyu-y) [inline device-plugin/ccnp-device-plugin/README.md:57]: According to testing, some correction here:
kubectl apply -k .
Also user have to change the imagePullPolicy to 'IfNotPresent'. Better to add note for this

**Review by Ruoyu-y — CHANGES_REQUESTED** [review]

File: .github/cspell.json [hunk 2]
```diff
@@ -9,6 +9,13 @@
...
     "dictionaries": [],
     "words": [
-        "ccnp"
+        "ccnp",
+        "TDVM",
+        "kustomization",
+        "srwxr",
+        "udev",
+        "udevadm",
+        "daemonset",
```
Comment (hwang37) [inline .github/cspell.json:18]: Small comment: please use alphabeta order for the words

File: device-plugin/ccnp-device-plugin/README.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+This will enable the services in the PODs to be able to communicate to the device for quote, measurements etc.
+And will also facilitate other CCNP requirements like mount certain directories to CCNP service PODs and workload PODs.
+
+## Introduction
+
+Currently, the CCNP device plugin has following capabilities:
+- expose tdx guest device node in TDVM to PODs
+- mount Unix Domain Socket dir /run/ccnp/uds into CCNP daemonset PODs and workload PODs to enable intra-node communication
+
+The CCNP device plugin will response to following resource request defined in POD definition yaml:
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:13]: response => respond

File: device-plugin/ccnp-device-plugin/README.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+```
+
+## Installation
+
+### Prerequisite
+The CCNP device plugin need to deploy on VM nodes with guest TEE devices(currently only TDX guest device is supported). So the deployment
+of the plugin daemonset is based on the node label set by [Node Feature Discovery](https://github.com/kubernetes-sigs/node-feature-discovery/).
+So we need to install the NFD and corresponding label rules.
+
+1. setup following udev rule to enable other user in the node to read and write to tdx guest device node
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:27]: read and write => read or write

File: device-plugin/ccnp-device-plugin/README.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+```
+
+> Note: if you are using containerd as the default runtime for kubernetes. Please remember to use the following command to import the image into containerd:
+```
+docker save -o ccnp-device-plugin.tar ccnp_device_plugin:0.1
+ctr -n=k8s.io image import ccnp-device-plugin.tar
+```
+
+### Deploy as DaemonSet
+Use bellow helm command to deploy:
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:87]: bellow => below

File: .github/cspell.json (hunk 2 above)
Comment (hwang37) [reply .github/cspell.json:18]: More words can be added in dictionary, for example,
(I'm not sure whether we need add both lowcase or uppercase words, please check the cspell.json rules)

daemonset
CCNP
ccnp
POD
NFD
tdx
TEE
VM

File: device-plugin/ccnp-device-plugin/README.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+> Note: you may need to edit settings in deploy/helm/ccnp-device-plugin/value.yaml according to you cluster status.
+```
+cd device-plugin/ccnp-device-plugin/
+helm install ccnp-device-plugin deploy/helm/ccnp-device-plugin
+
+```
+
+After the deployment, for TDVM node, you can see bellow resource info:
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:95]: bellow => below

File: .github/cspell.json (hunk 2 above)
Comment (hairongchen) [reply .github/cspell.json:18]: for more words, let's add and merge when our PRs get's in and the case is fine as per my check result

File: device-plugin/ccnp-device-plugin/README.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+The CCNP device plugin is based on Kubernetes plugin framework to expose host's TEE devices and other required resources to PODs.
+This will enable the services in the PODs to be able to communicate to the device for quote, measurements etc.
+And will also facilitate other CCNP requirements like mount certain directories to CCNP service PODs and workload PODs.
+
+## Introduction
+
+Currently, the CCNP device plugin has following capabilities:
+- expose tdx guest device node in TDVM to PODs
+- mount Unix Domain Socket dir /run/ccnp/uds into CCNP daemonset PODs and workload PODs to enable intra-node communication
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:11]: Unix?
=> Change to another word? Linux? Network?...

File: .github/cspell.json [hunk 8]
```diff
@@ -9,6 +9,13 @@
     "dictionarydefinitions": [],
     "dictionaries": [],
     "words": [
-        "ccnp"
+        "ccnp",
```
Comment (hwang37) [inline .github/cspell.json:12]: add "NFD"

File: device-plugin/ccnp-device-plugin/README.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+2. prepare the shared Unix Domain Socket directory to be mounted to both ccnp service pods and workload pods
+```
+mkdir -p /run/ccnp/uds
+chmod o+w /run/ccnp/uds
+```
+
+3. deploy NFD
+
+> Note: when node-feature-discovery new [release v0.14](https://github.com/kubernetes-sigs/node-feature-discovery/issues/1250) is ready, below command can be used to deploy NFD with TDVM support:
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:46]: new???

File: device-plugin/ccnp-device-plugin/README.md [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+```
+kubectl apply -k https://github.com/kubernetes-sigs/node-feature-discovery/deployment/overlays/default?ref=v0.14
+```
+
+before that, please build own image and deploy:
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:52]: This is step by step guide, so the steps should be sequential, no need to use "before that..." 

For example, how about adjusting the paragraph order by put "before that,..." paragraph before "kubectrl apply..." part.

File: device-plugin/ccnp-device-plugin/README.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+After deployment, following label can be found in the VM node:
+```
+kubectl get node -o json | jq .items[].metadata.labels | grep tdx-guest
+  "intel.feature.node.kubernetes.io/tdx-guest": "enabled",
+```
+Above label can be used as node selector by CCNP device plugin daemonset and CCNP services daemonset.
+
+
+### Build docker image
+The Dockerfile for the service can be found under device-plugin/ccnp-device-plugin/container directory.
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:75]: Format comment:
change to hyper URL format?

File: device-plugin/ccnp-device-plugin/README.md [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+
+### Build docker image
+The Dockerfile for the service can be found under device-plugin/ccnp-device-plugin/container directory. 
+Use the following command to build the image:
+```
+docker build -t ccnp_device_plugin:0.1 -f device-plugin/ccnp-device-plugin/container/Dockerfile .
+```
+
+> Note: if you are using containerd as the default runtime for kubernetes. Please remember to use the following command to import the image into containerd:
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:81]: if you are using containerd as the default runtime for kubernetes, don't forget to use the following commands to import the image into containerd first:

File: device-plugin/ccnp-device-plugin/README.md [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+  --------                 --------     ------
+  cpu                      1250m (15%)  600m (7%)
+  memory                   510Mi (6%)   690Mi (9%)
+...
+  tdx.intel.com/tdx-guest  0            0
+...
+```
+
+### Testing
+One can deploy a CCNP quote service with tdx-guest resource request in the DaemonSet definition yaml:
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:125]: One => User

File: device-plugin/ccnp-device-plugin/README.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+```
+...
+        resources:
+          limits:
+            tdx.intel.com/tdx-guest: 1
+...
+```
+
+And after the quote server POD is started, follow resource and directory can be found in the container of the POD:
```
Comment (hwang37) [inline device-plugin/ccnp-device-plugin/README.md:134]: follow => the following

File: device-plugin/ccnp-device-plugin/README.md (hunk 9 above)
Comment (hairongchen) [reply device-plugin/ccnp-device-plugin/README.md:46]: we have a dependency commit merged after last release, so before this new release is ready, user need to build NFD image themselves to sue NFD feature for TDVM

File: device-plugin/ccnp-device-plugin/README.md (hunk 10 above)
Comment (hairongchen) [reply device-plugin/ccnp-device-plugin/README.md:52]: change the paragraph sequence

File: device-plugin/ccnp-device-plugin/README.md (hunk 14 above)
Comment (hairongchen) [reply device-plugin/ccnp-device-plugin/README.md:134]: updated, thanks!

File: device-plugin/ccnp-device-plugin/README.md (hunk 12 above)
Comment (hairongchen) [reply device-plugin/ccnp-device-plugin/README.md:81]: updated, thanks!

File: device-plugin/ccnp-device-plugin/README.md (hunk 11 above)
Comment (hairongchen) [reply device-plugin/ccnp-device-plugin/README.md:75]: updated link format

**Review by kenplusplus — APPROVED** [review]
LGTM

File: .github/cspell.json [hunk 15]
```diff
@@ -15,6 +15,13 @@
     "words": [
         "ccnp",
         "containerd",
+        "daemonset",
+        "kustomization",
+        "NFD",
+        "srwxr",
```
Comment (Ruoyu-y) [inline .github/cspell.json:21]: Not needed. As it is part of code block.

File: .github/cspell.json [hunk 16]
```diff
@@ -15,6 +15,13 @@
...
+        "TDVM",
+        "udev",
+        "udevadm"
```
Comment (Ruoyu-y) [inline .github/cspell.json:24]: Not needed. As it is part of code block

File: .github/cspell.json [hunk 17]
```diff
@@ -15,6 +15,13 @@
...
         "Eventlog",
         "grpcurl",
         "proto",
```
Comment (Ruoyu-y) [inline .github/cspell.json:27]: Same issue on the sequence

**Review by Ruoyu-y — APPROVED** [review]

**Review by Ruoyu-y — APPROVED** [review]
LGTM

**Pull request merged**